### PR TITLE
FIX: Split BadgeGrant in a job for each badge

### PR DESCRIPTION
**Pull request by nbianca**
This commit splits execution of the `BadgeGrant` job in smaller chunks,
enqueuing a job for each badge, one to enqueue the other jobs and one to
ensure consistency.

To ensure that the "ensure consistency" job runs only once after all the
other jobs have been enqueued, a Redis counter is used.

### CODE REVIEW

File: app/jobs/regular/ensure_badge_consistency.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module Jobs
+  class EnsureBadgeConsistency < ::Jobs::Base
```
Comment (nbianca) [inline app/jobs/regular/ensure_badge_consistency.rb:4]: Should I add this here?

```
sidekiq_options queue: "low"
```

File: app/jobs/regular/grant_badge.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+module Jobs
+  class GrantBadge < ::Jobs::Base
```
Comment (nbianca) [inline app/jobs/regular/grant_badge.rb:4]: Should I add this here?

```
sidekiq_options queue: "low"
```

File: app/jobs/regular/grant_badge.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    def execute(args)
+      return unless SiteSetting.enable_badges
+
+      badge = Badge.enabled.find_by(id: args[:badge_id])
+      return unless badge
+
+      # Cancels the scheduled job to ensure badge consistency as the badges are
```
Comment (tgxworld) [inline app/jobs/regular/grant_badge.rb:11]: I don't think this is sufficient since Sidekiq jobs can be processed concurrently. The `Jobs::EnsureBadgeConsistency` job could already be running. If you want to ensure that stuff in the `ensure_badge_consistency` job is not running while this job is running, you will have to use some sort of mutex.

File: app/jobs/regular/grant_badge.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      # mutating during `BadgeGranter.backfill`.
+      Jobs.cancel_scheduled_job(:ensure_badge_consistency)
```
Comment (tgxworld) [inline app/jobs/regular/grant_badge.rb:13]: Is `Jobs:EnsureBadgeConsistency` supposed to be a scheduled job because it is written as a regular job at the moment.

File: app/jobs/regular/grant_badge.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      # If this instance is among the last few jobs to be processed, consider
+      # rescheduling the EnsureBadgeConsistency job. This ensures the job is
+      # scheduled only once after all badges scheduled by GrantAllBadges have
+      # been granted.
+      if Sidekiq::Queue.new.count { |job| job.klass =~ /GrantBadge/ } == 0
+        DistributedMutex.synchronize("ensure_badge_consistency") do
+          Jobs.cancel_scheduled_job(:ensure_badge_consistency)
+          Jobs.enqueue_in(1.minute, :ensure_badge_consistency)
+        end
+      end
```
Comment (tgxworld) [inline app/jobs/regular/grant_badge.rb:34]: I'm curious what problems we are seeing so that we need to split this out into another job.

File: app/jobs/regular/grant_badge.rb (hunk 3 above)
Comment (tgxworld) [reply app/jobs/regular/grant_badge.rb:11]: O I see a distributed mutex is being used below.

File: app/jobs/regular/grant_badge.rb (hunk 5 above)
Comment (tgxworld) [inline app/jobs/regular/grant_badge.rb:34]: I find this hard to reason about and am not even sure if it will work. What we want here is something like [Sidekiq Pro's batch feature](https://github.com/sidekiq/sidekiq/wiki/Batches). Can you see if you can support something similar? Basically, we want an API to enqueue a bunch of jobs, monitor those jobs for completion and then schedule a job after all jobs have exited.

File: app/jobs/regular/ensure_badge_consistency.rb (hunk 1 above)
Comment (tgxworld) [reply app/jobs/regular/ensure_badge_consistency.rb:4]: Is this job blocking the default queue frequently? Otherwise, I wouldn't worry about the queue it is in for now.

Comment (nbianca) [reply app/jobs/regular/ensure_badge_consistency.rb:4]: The old job was running for more than 2h. I think a considerable amount of the time was spent doing this, so yes, I believe it may block a thread of a Sidekiq process until the job is over. However, I do not think our queue handling is optimal right now so this line might not have the desired effect.

File: app/jobs/regular/grant_badge.rb (hunk 5 above)
Comment (nbianca) [reply app/jobs/regular/grant_badge.rb:34]: In my tests, it works, but there are a few edge cases, for example jobs are spread and it takes more than 1 minute for the next job "in batch" to run. In this case, the call to `cancel_scheduled_job` is too late.

**Comment by nbianca** [comment]
@tgxworld, I have changed the approach. What do you believe about this one?

File: app/jobs/regular/grant_badge.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        )
+      end
+
+      # If this instance is the last job to be processed, schedule the
+      # EnsureBadgeConsistency job. This guarantees it runs only once after all
+      # badges have been granted.
+      if Discourse.redis.decr("grant_badge_remaining") <= 0
+        Discourse.redis.del("grant_badge_remaining")
+        Jobs.enqueue(:ensure_badge_consistency)
+      end
```
Comment (tgxworld) [inline app/jobs/regular/grant_badge.rb:27]: I think this will work but I'm still not a fan of this approach because we are arguably introducing some sort of coupling between two jobs.

**Comment by tgxworld** [comment]
> I have not found any implementation details on Sidekiq's batch feature.

Let's try to break the feature down. 

* The scheduled job enqueues the `grant_badge` job and we can collect the ids of the jobs. 
* With the job ids, we can pass them to a thread that polls Sidekiq at regular intervals to check if the jobs have been completed. 
* On completion, we run a callback or enqueue a job.

File: app/jobs/base.rb [hunk 7]
```diff
@@ -461,6 +461,45 @@ def self.enqueue_at(datetime, job_name, opts = {})
     enqueue_in(secs, job_name, opts)
   end
 
+  @@enqueue_after = []
```
Comment (tgxworld) [inline app/jobs/base.rb:464]: I don't think we should hold the callback in memory because a process restart will cause the callback to be lost.

Comment (tgxworld) [reply app/jobs/base.rb:464]: Redis seems like a good storage option here.

File: app/jobs/base.rb [hunk 8]
```diff
@@ -461,6 +461,45 @@ def self.enqueue_at(datetime, job_name, opts = {})
...
+
+  def self.process_enqueue_after
+    @@enqueue_after.filter! do |job_ids, job_name, opts|
+      pending_job_ids = Set.new
+
+      Sidekiq::Queue.all.each do |queue|
+        queue.each { |job| pending_job_ids << job.jid if job_ids.include?(job.jid) }
+      end
```
Comment (tgxworld) [inline app/jobs/base.rb:472]: I don't think we can rely on `Sidekiq::Queue` because the number of jobs in the queue may be huge. Do have a read through https://www.mikeperham.com/2021/04/20/a-tour-of-the-sidekiq-api/ for more information. 

Can you see if we can use `Sidekiq::ScheduledSet.new` or `Sidekiq::RetrySet.new` to scan for the relevant jobs instead? That should be way more efficient and accurate.

File: app/jobs/base.rb (hunk 7 above)
Comment (tgxworld) [reply app/jobs/base.rb:464]: Do we also need to think about timeouts here to account for the case where jobs don't complete within the given timeframe? We don't want the callback to linger around forever.

Comment (nbianca) [reply app/jobs/base.rb]: I totally agree. I was working on this between holidays and did not have the chance to wrap it up. I was also not very happy with relying on a thread to perform the checks. I think the new approach is much better.

File: app/jobs/base.rb (hunk 8 above)
Comment (nbianca) [reply app/jobs/base.rb]: Yes, that's not ideal and I did not like it either. I find the alternatives just as bad.

I think the new approach is much better though.

File: app/jobs/base.rb [hunk 9]
```diff
@@ -461,6 +470,50 @@ def self.enqueue_at(datetime, job_name, opts = {})
     enqueue_in(secs, job_name, opts)
   end
 
+  def self.decrease_batch_count(batch_id)
```
Comment (tgxworld) [inline app/jobs/base.rb:473]: Let's mark this as private or at least `:nodoc:` so that it is obvious this isn't a public API.

File: app/jobs/base.rb [hunk 10]
```diff
@@ -461,6 +470,50 @@ def self.enqueue_at(datetime, job_name, opts = {})
...
+
+        # An instance of the same job might run several times (e.g. if it fails
+        # and is retried). When that happens, `batch` might be undefined because
+        # the batch job has already been enqueued and deleted from Redis.
+        Jobs.enqueue(batch["job_name"], JSON.parse(batch["opts"])) if batch["job_name"]
+      end
+    end
+  end
+
+  def self.enqueue_after(job_name, opts = {})
```
Comment (tgxworld) [inline app/jobs/base.rb:489]: Let's write some docs for this describing how this should be used and provide examples of its usage.

File: app/jobs/base.rb [hunk 11]
```diff
@@ -352,6 +354,8 @@ def perform(*args)
         self.class.clear_cluster_concurrency_lock!
       end
 
+      Jobs.decrease_batch_count(batch_id) if batch_id
```
Comment (tgxworld) [inline app/jobs/base.rb:357]: What if a job never completes successfully and exhausts the maximum number of retries? Do we want to handle cases like this or do we just put a giant warning in the API that all errors in the jobs that are enqueued must be rescued in order for the "callback" job to be enqueued?

File: app/jobs/base.rb [hunk 12]
```diff
@@ -461,6 +470,50 @@ def self.enqueue_at(datetime, job_name, opts = {})
...
+
+    raise "Nested enqueue_after is not supported" if Thread.current[:current_batch_id]
+
+    Thread.current[:current_batch_id] = batch_id = SecureRandom.hex(12)
+
+    Sidekiq.redis do |redis|
+      # Jobs in the batch might not be enqueued fast enough and some might
+      # finish before the batch is fully enqueued. Starting with a ref_count of
+      # 1 and decrementing it after all jobs are enqueued ensures that all batch
+      # jobs run.
```
Comment (tgxworld) [inline app/jobs/base.rb:504]: I don't quite understand the comment here. Will you be able to provide a simple example of why we need to do this?

File: app/jobs/base.rb (hunk 11 above)
Comment (nbianca) [reply app/jobs/base.rb:357]: A success or failure is treated the same way.

File: app/jobs/base.rb (hunk 12 above)
Comment (nbianca) [reply app/jobs/base.rb:504]: Sure.

Let's consider the following block:

```ruby
Jobs.enqueue_after(:a_job) do
  Jobs.enqueue(:some_other_job)
  sleep 3
  Jobs.enqueue(:some_other_job)
end
```

Let's also consider that the initial `ref_count` is 0, not 1 like in this PR.

`Jobs.enqueue_after` will set the `ref_count` to `0` initially. `Jobs.enqueue` will increase the `ref_count` to 1. Then this thread will sleep 3 seconds, but Sidekiq will continue to run and might execute `some_other_job`. If it does that, `ref_count` will decrease to `0` and `a_job` will be enqueued.

Let's consider this implementation with the initial `ref_count` is `1`.

`Jobs.enqueue_after` sets `ref_count` to `1`. `some_other_job` will be enqueued and increase `ref_count` to `2`. `sleep 3` will allow for the job to run and decrease `ref_count` back to 1. Then it goes back up to `2` because of the other `Jobs.enqueue`. Then, the rest of `enqueue_after` and Sidekiq will run in parallel, so either of them will decrease `ref_count` to 1 and the other one down to `0` and finally enqueue `a_job`.

File: app/jobs/base.rb (hunk 11 above)
Comment (tgxworld) [reply app/jobs/base.rb:357]: What is our definition of failure here? Is it the job exhausting the number of retired and ending up in the dead queue or is it just the job failing once?

**Comment by nbianca** [comment]
I am closing this in favor of #30945.
